### DISCUSSION TITLE
RFC: Revamp identification of TCP/UDP port bind and unbind events

### DISCUSSION THREAD

**Discussion by balajiv113** [lima-vm/lima#2586] [Ideas]
**Problem**
Major problems with current approach are,
- Polling to find the open or closed ports using /proc/net/{tcp / udp / tcp6 / udp6}
- This model doesn't work for containers (As they might have iptable rules to bind things differently)

**Possible Solution** 
**1. Using eBPF program (Recommended)**
In this approach we make use of eBPF program https://github.com/balajiv113/trackport written in go for identifying tcp and udp opening and closing.

Pros
- Easy to integrate and maintain as only minimal code is maintained in c and also has go interoperability
- Support all events tcp, udp, tcp6, udp6 open and close of ports

Cons
- Events are published still as asyn (eBPF write to a map and go prog reads from that map). But this will be lot faster compared to 2 sec polling
- Events used by us is not supported in all flavours of linux. Based on latest testing Alpine and Archlinux doesn't support the required events (https://github.com/balajiv113/lima/pull/3/checks)

**2. Using go-libaudit**
In this approach we make use of go-libaudit that we already use in guest agent to identify port opening and closing.

Pros
- Supported in all linux flavours

Cons
- Unable to identify port closing using syscalls (When program is force terminated).
- As we are reading from logs and based on syscall events, at times it can be slower

Creating this RFC mainly to brainstorm ideas as we don't have a single drop in replacement with either approach. But it nearly solves the problem

Some further ideas to explore are 
- Trying with combination of eBPF and go-libaudit
- Analyse further why alpine doesn't support eBPF events
- Custom kernel extension (Mostly will try to avoid, as worst case we can explore it)

**Comment by Nino-K** [comment]
@balajiv113, we are using [this](https://github.com/rancher-sandbox/rancher-desktop-agent) component in Rancher Desktop. Why couldn't Lima adopt a similar approach since it doesn't require polling? As I mentioned, I’m prepared to make the necessary adjustments to the project to make it compatible with Lima.

**Reply by balajiv113** [reply]
@Nino-K The problem with that is, for each container system we need to implement it using their respective API.

This has lots of drawbacks like,
- Too many unwanted dependencies 
- Maintainability
- If a newer system comes in, need to do implementation again

The proposal is to derive a generic solution across the board. This way we will never need to worry about any new container system.

> doesn't require polling?

Here with polling what i menat is, bpf program polls the kerner ring buffer from userspace program. This is not done with some intervals its almost instantaneous

**Comment by AkihiroSuda** [comment]
> Trying with combination of eBPF and go-libaudit

👍 I think this is the most realistic approach

> Analyse further why alpine doesn't support eBPF events

👍 

> Custom kernel extension (Mostly will try to avoid, as worst case we can explore it)

👎 
It is quite hard to maintain kernel modules for multiple distros.
This is also likely to affect stability.

**Reply by balajiv113** [reply]
Able to come up with a approach inspired mostly by datadog-agent repo. Using go-libaudit was complex as it didn't provide much data mainly during close operations.

Created a mix of fentry vs kprobe. If fentry is present it will take fentry else will fallback to kprobe. Kprobe is supported almost everywhere. Ran CI job to verify it

**Reply by balajiv113** [reply]
Results are consistent across all distro. Will fix the CI for failing ones. I think this should work out for all usecases

**Reply by balajiv113** [reply]
Seeing a issue with nerdctl it doesn't bind on host, only routed via iptables so there is no trace of it via ebpf binds events (Not even present in /proc/tcp tables). Need to identify a way to support this.

For some distros its working as per CI but failing in alpine need to check on this.

**Comment by balajiv113** [comment]
> Analyse further why alpine doesn't support eBPF events

Required option CONFIG_FUNCTION_TRACER not set
Required option CONFIG_DYNAMIC_FTRACE not set

These 2 options are missing in alpine due to which events are not published

**Reply by balajiv113** [reply]
Archlinux is missing the following

Required option CONFIG_BPF_EVENTS not set
Required option CONFIG_FTRACE_SYSCALLS not set
Required option CONFIG_FUNCTION_TRACER not set
Required option CONFIG_DYNAMIC_FTRACE not set
Required option CONFIG_KPROBE_EVENTS not set
Required option CONFIG_UPROBES not set
Required option CONFIG_UPROBE_EVENTS not set
